modules/apps/frontend-data-set/frontend-data-set-web/src/main/resources/META-INF/resources/management_bar/components/filters/implementation/clientExtensionFilter.tsx

```diff
@@ -0,0 +1,188 @@
+/**
+ * SPDX-FileCopyrightText: (c) 2000 Liferay, Inc. https://liferay.com
+ * SPDX-License-Identifier: LGPL-2.1-or-later OR LicenseRef-Liferay-DXP-EULA-2.0.0-2023-06
+ */
+
+import {ClientExtension} from 'frontend-js-components-web';
+import PropTypes from 'prop-types';
+import React, {useEffect, useState} from 'react';
+
+import type {FDSFilter} from '@liferay/js-api/data-set';
+
+import type {
+	FilterImplementation,
+	FilterImplementationArgs,
+	SetFilterArgs,
+} from '../Filter';
+
+export interface ClientExtensionFilterImplementationArgs
+	extends FilterImplementationArgs<unknown> {
+	cxFilterImplementation?: FDSFilter<unknown>;
+	cxFilterURL: string;
+}
+
+function getSelectedItemsLabel({
+	cxFilterImplementation,
+	cxFilterURL,
+	selectedData,
+}: ClientExtensionFilterImplementationArgs) {
+	if (!cxFilterImplementation) {
+		return '...';
+	}
+
+	if (!cxFilterImplementation.descriptionBuilder) {
+		console.warn(
+			'The filter client extension',
+			cxFilterURL,
+			'is not exporting a descriptionBuilder() function,',
+			'thus the filter description will not show meaningful information.',
+			'The client extension should be reworked.'
+		);
+
+		return '...';
+	}
+
+	try {
+		return cxFilterImplementation.descriptionBuilder(selectedData);
+	}
+	catch (error) {
+		console.error(
+			'The filter client extension',
+			cxFilterURL,
+			'caused an error when trying to render the filter description in',
+			'the descriptionBuilder<() function.',
+			'The client extension needs to be fixed.',
+			error
+		);
+
+		return '...';
+	}
+}
+
+function getOdataString({
+	cxFilterImplementation,
+	cxFilterURL,
+	selectedData,
+}: ClientExtensionFilterImplementationArgs) {
+	if (!cxFilterImplementation) {
+		return '';
+	}
+
+	if (!cxFilterImplementation.oDataQueryBuilder) {
+		console.error(
+			'The filter client extension',
+			cxFilterURL,
+			'is not exporting a oDataQueryBuilder() function,',
+			'thus the filter will NOT work.',
+			'The client extension needs to be fixed.'
+		);
+
+		return '';
+	}
+
+	try {
+		return cxFilterImplementation.oDataQueryBuilder(selectedData);
+	}
+	catch (error) {
+		console.error(
+			'The filter client extension',
+			cxFilterURL,
+			'caused an error when trying to compute the filter query in the',
+			'oDataQueryBuilder() function.',
+			'The client extension needs to be fixed.',
+			error
+		);
+
+		return '';
+	}
+}
+
+function spinnerHTMLElementBuilder(): HTMLElement {
```

Comment:
Can we use `ClayLoadingIndicator` instead? Something like
```js
{cxFilterImplementationIsGood ? : <ClientExtension /> <ClayLoadingIndicator> }
```

Maybe this should be left for parent component to handle, wherever the place is that resolves the implementation.

Reply:
Oh, that's true. I thought I couldn't use it because the function is returning a raw HTML element, but reflecting now on it, of course I can... :sweat_smile:

Reply:
> Maybe this should be left for parent component to handle,

Nope, because the parent of `ClientExtensionFilter` if `Filter` and I don't want to place any filter specific logic there if possible. Otherwise we break the `FilterImplementation` abstraction :grimacing: .

But I think I can simply use it in the function itself.

Reply:
Nope. I've tried to use the React component and it is not possible, because the `spinnerHTMLElementBuilder` function ultimately is passed to the [ClientExtension](https://github.com/liferay/liferay-portal/blob/0d464205039fc0645e49ea51687bac0b12e5af68/modules/apps/frontend-js/frontend-js-components-web/src/main/resources/META-INF/resources/ClientExtension.tsx) component that is expecting a function returning an `HTMLElement`, not a React element.

That is because CXs deal with raw DOM elements since they are not forced to use React and this `spinnerHTMLElementBuilder` is like a fallback CX implementation (even though it is inside the portal).

We could "fix" this by adding some ifs and making the `ClientExtension` more complex but I doubt the trade off would be worth the effort. IMO the resulting code would be less maintainable... :thinking:

Reply:
I could also use `ReactDOM.render` to render a `ClayLoadingIndicator` and return it as an `HTMLElement` but I'm not sure what's better: if that, or using Clay's markup directly :shrug: . You are the expert in this, I'll do what you decide...

Reply:
> We could "fix" this by adding some ifs and making the ClientExtension more complex but I doubt the trade off would be worth the effort. IMO the resulting code would be less maintainable...

That actually doesn't sound too bad to me. We already have there a `useEffect` that triggers on `htmlElementBuilder`. I believe we only only need to change the return line, to something like:
```ts
return htmlElementBuilder ? <div ref={containerRef}></div> : <ClayLoadingIndicator/>;
```

Reply:
Oh, nice solution :+1: 

And it works 🤗

Reply:
done